src/front/wgsl/lower/mod.rs

```diff
@@ -0,0 +1,2368 @@
+    /// we're lowering.
+    ///
+    /// [`TranslationUnit::expressions`]: ast::TranslationUnit::expressions
+    ast_expressions: &'temp Arena<ast::Expression<'source>>,
+
+    /// A reference to [`TranslationUnit::types`] for the translation unit
+    /// we're lowering.
+    ///
+    /// [`TranslationUnit::types`]: ast::TranslationUnit::types
+    types: &'temp Arena<ast::Type<'source>>,
+
+    // Naga IR values.
+    /// The map from the names of module-scope declarations to the Naga IR
+    /// `Handle`s we have built for them, owned by `Lowerer::lower`.
+    globals: &'temp mut FastHashMap<&'source str, LoweredGlobalDecl>,
+
+    /// A map from `ast::Local` handles to the Naga expressions we've built for them.
+    ///
+    /// The Naga expressions are either [`LocalVariable`] or
+    /// [`FunctionArgument`] expressions.
+    ///
+    /// [`LocalVariable`]: crate::Expression::LocalVariable
+    /// [`FunctionArgument`]: crate::Expression::FunctionArgument
+    local_table: &'temp mut FastHashMap<Handle<ast::Local>, TypedExpression>,
+
+    typifier: &'temp mut Typifier,
+    variables: &'out mut Arena<crate::LocalVariable>,
+    naga_expressions: &'out mut Arena<crate::Expression>,
+    named_expressions: &'out mut FastHashMap<Handle<crate::Expression>, String>,
+    arguments: &'out [crate::FunctionArgument],
+    module: &'out mut crate::Module,
+}
+
+impl<'a, 'temp> StatementContext<'a, 'temp, '_> {
+    fn reborrow(&mut self) -> StatementContext<'a, '_, '_> {
+        StatementContext {
+            local_table: self.local_table,
+            globals: self.globals,
+            types: self.types,
+            ast_expressions: self.ast_expressions,
+            typifier: self.typifier,
+            variables: self.variables,
+            naga_expressions: self.naga_expressions,
+            named_expressions: self.named_expressions,
+            arguments: self.arguments,
+            module: self.module,
+        }
+    }
+
+    fn as_expression<'t>(
+        &'t mut self,
+        block: &'t mut crate::Block,
+        emitter: &'t mut Emitter,
+    ) -> ExpressionContext<'a, 't, '_>
+    where
+        'temp: 't,
+    {
+        ExpressionContext {
+            local_table: self.local_table,
+            globals: self.globals,
+            types: self.types,
+            ast_expressions: self.ast_expressions,
+            typifier: self.typifier,
+            naga_expressions: self.naga_expressions,
+            module: self.module,
+            local_vars: self.variables,
+            arguments: self.arguments,
+            block,
+            emitter,
+        }
+    }
+
+    fn as_output(&mut self) -> OutputContext<'a, '_, '_> {
+        OutputContext {
+            ast_expressions: self.ast_expressions,
+            globals: self.globals,
+            types: self.types,
+            module: self.module,
+        }
+    }
+}
+
+pub struct ExpressionContext<'source, 'temp, 'out> {
```

Comment:
Might be worth adding a comment:

```suggestion
/// State for lowering an `ast::Expression` to Naga IR.
///
/// Not to be confused with `parser::ExpressionContext`.
pub struct ExpressionContext<'source, 'temp, 'out> {
```